Frameworks/CoreImage/CIImage.mm

```diff
@@ -73,90 +74,164 @@ @implementation CIImage
 /**
  @Status Interoperable
 */
-- (CIImage*)init {
++ (CIImage*)emptyImage {
+    return [[[CIImage alloc] init] autorelease];
+}
+
+/**
+ @Status Interoperable
+*/
+- (CGRect)extent {
+    return self->_extent;
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)init {
     if (self = [super init]) {
-        _cgImage = nil;
-        _color = nil;
-        _filter = nil;
         _extent = CGRectInfinite;
     }
     return self;
 }
 
+- (instancetype)_initWithCGImage:(CGImageRef)image color:(CIColor*)color {
+    if (self = [self initWithCGImage:image]) {
+        _color.attach([static_cast<id>(color) copy]);
+    }
+
+    return self;
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithColor:(CIColor*)color {
+    if (!color) {
+        [self release];
+        return nil;
+    }
+
+    if (self = [self init]) {
+        _color.attach([static_cast<id>(color) copy]);
+    }
+
+    return self;
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithCGImage:(CGImageRef)cgImage {
+    if (!cgImage) {
+        [self release];
+        return nil;
+    }
+
+    if (self = [self init]) {
+        _cgImage.reset(CGImageRetain(cgImage));
+        _extent = CGRectMake(0, 0, CGImageGetWidth(cgImage), CGImageGetHeight(cgImage));
+    }
+
+    return self;
+}
+
 /**
  @Status Stub
  @Notes
 */
-+ (CIImage*)emptyImage {
+- (instancetype)initWithCGImage:(CGImageRef)image options:(NSDictionary*)d {
     UNIMPLEMENTED();
     return StubReturn();
 }
 
 /**
  @Status Interoperable
 */
-+ (CIImage*)imageWithColor:(CIColor*)color {
-    CIImage* ret = [[CIImage alloc] init];
-    if (ret != nil) {
-        ret->_color.attach([static_cast<id>(color) copy]);
-        ret->_extent = CGRectMake(INFINITY, INFINITY, CGFLOAT_MAX, CGFLOAT_MAX);
+- (instancetype)initWithImage:(UIImage*)image {
+    if (!image) {
+        [self release];
+        return nil;
     }
-    return [ret autorelease];
+    return [self initWithCGImage:[image CGImage]];
 }
 
 /**
  @Status Stub
  @Notes
 */
-+ (CIImage*)imageWithBitmapData:(NSData*)d bytesPerRow:(size_t)bpr size:(CGSize)size format:(CIFormat)f colorSpace:(CGColorSpaceRef)cs {
+- (instancetype)initWithImage:(UIImage*)image options:(NSDictionary*)options {
     UNIMPLEMENTED();
     return StubReturn();
 }
 
 /**
  @Status Interoperable
 */
-+ (CIImage*)imageWithCGImage:(CGImageRef)image {
-    CIImage* ret = [[CIImage alloc] init];
-    if (ret != nil) {
-        ret->_cgImage.attach([static_cast<id>(image) copy]);
-        ret->_extent = CGRectMake(0, 0, CGImageGetWidth(image), CGImageGetHeight(image));
+- (instancetype)initWithData:(NSData*)data {
+    if (!data) {
+        [self release];
+        return nil;
     }
-    return [ret autorelease];
+    return [self initWithCGImage:_CGImageGetImageFromData((void*)[data bytes], [data length])];
```

Comment:
sadly no they have ways to load only JPEG and PNG, it if was any other format we have to go with this.

---
In reply to: [95705644](https://github.com/Microsoft/WinObjC/pull/1625#discussion_r95705644) [](ancestors = 95705644)